### ISSUE TITLE
REST API metadata/callables endpoint returns empty response for some callables that do exist in the DB

### ISSUE THREAD

**Issue by MagielBruntink** [fasten-project/fasten#253]
## Describe the bug
For some callables the REST API endpoint metadata/callables returns an empty result, as if the callables do not exist in the DB. We verified that indeed they are present in the DB (see below). In some test runs with large numbers of callables, this occurs in about 40-50% of the cases, leading to a large reduction in functionality in for instance the Maven Plugin, which relies on this endpoint. Tagging @tmortagne here for awareness.

Example callable for which there is unexpected behavior:
```
curl -X POST -H "Content-Type: application/json" -d '["fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject"]' https://api.fasten-project.eu/api/metadata/callables?allAttributes=true
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   181  100     2  100   179     19   1754 --:--:-- --:--:-- --:--:--  1774
{}
```
We do not expect the empty result here {}, as the callable is in the DB. The result should be:
```
{"fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject" : {"access":"public","defined":true}}
```

An example from the same package, which works fine:
```
curl -X POST -H "Content-Type: application/json" -d '["fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.hash/FarmHashFingerprint64.hashBytes(%2Fjava.lang%2FByteType%5B%5D,%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FIntegerType)HashCode"]' https://api.fasten-project.eu/api/metadata/callables?allAttributes=true
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   427  100   231  100   196   2158   1831 --:--:-- --:--:-- --:--:--  3990
{"fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.hash/FarmHashFingerprint64.hashBytes(%2Fjava.lang%2FByteType%5B%5D,%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FIntegerType)HashCode":{"access":"public","defined":true}}
```

## Are they really in the DB?
Yes. We found the global ID of the SynchronizedList.set callable above and with the following POST request you can see the callable is indeed in the DB with the expected URI.
```
curl -X POST -H "Content-Type: application/json" -d '[1295364849]' https://api.fasten-project.eu/api/callable_uris
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   204  100   192  100    12    462     28 --:--:-- --:--:-- --:--:--   491
{"1295364849":"fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject"}
```

Its full entry is:
```
{
    "id": 1295364849,
    "module_id": 153977054,
    "fasten_uri": "/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject",
    "is_internal_call": true,
    "created_at": null,
    "line_start": 417,
    "line_end": 419,
    "metadata": {"access": "public", "defined": true}
  }
```

## Logs
From a local test run with the REST API, the following log info is observed:
```
2021-03-10 08:37:54.381 DEBUG 1 --- [io-8080-exec-10] m.m.a.RequestResponseBodyMethodProcessor : Read "application/json;charset=UTF-8" to [[fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedLis (truncated)...]
2021-03-10 08:37:54.395 DEBUG 1 --- [io-8080-exec-10] o.s.w.s.m.m.a.HttpEntityMethodProcessor : Using 'application/json', given [*/*] and supported [application/json]
2021-03-10 08:37:54.395 DEBUG 1 --- [io-8080-exec-10] o.s.w.s.m.m.a.HttpEntityMethodProcessor : Writing ["{}"]
2021-03-10 08:37:54.396 DEBUG 1 --- [io-8080-exec-10] o.s.web.servlet.DispatcherServlet : Completed 200 OK
```

**Comment by tmortagne** [comment]
>  at least the default metadata {"access":"public","defined":true},

Note that the Maven plugin does not request all metadata, but only those listed by the various analyzers (through `RiskAnalyzer#getMetadatas()`), so maybe that's your issue ?

**Comment by MagielBruntink** [comment]
> > at least the default metadata {"access":"public","defined":true},
> 
> Note that the Maven plugin does not request all metadata (the idea behind it being to reduce to the minimum required the data stored in memory and possibly the processing time on server side), but only those listed by the various analyzers (through `RiskAnalyzer#getMetadatas()`), so maybe that's your issue ?

No, that's not the issue. Also in the examples above the allAttributes=true parameter is set, still the result is totally empty for the Synchronized$SynchronizedList.set example, and many others. In case the requested metadata attributes are empty, the result should be:
`{"fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject" : {}`, not `{}`.

**Comment by MagielBruntink** [comment]
@MihhailSokolov I think the problem is here: https://github.com/fasten-project/fasten/blob/5fa7ae43794d8df6311898b65b39c3b3a5f65981/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/impl/StitchingApiServiceImpl.java#L81 where the full FASTEN URI is split on "$", which is an error because the "$" also appears later as separator for inner classes, for example in the problematic URI `fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject`. The split goes bad and the DB query returns no result.

**Comment by tmortagne** [comment]
> No, that's not the issue.

OK I must have read your first post too quickly.

**Comment by MihhailSokolov** [comment]
@MagielBruntink Turns out the problem was in parsing the full FASTEN URI in `FastenUriUtils.parseFullFastenUri(...)`. I fixed it in 66e6b71 
Your request now returns the proper result:
```
curl -X POST -H "Content-Type: application/json" -d '["fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject"]' https://api.fasten-project.eu/api/metadata/callables?allAttributes=true
{"fasten://mvn!com.google.guava:guava$24.1-jre/com.google.common.collect/Synchronized$SynchronizedList.set(%2Fjava.lang%2FIntegerType,%2Fjava.lang%2FObject)%2Fjava.lang%2FObject":{"access":"public","defined":true}}
```

**Comment by MagielBruntink** [comment]
Thank you @MihhailSokolov for the quick fix! I can confirm that the endpoint is now working properly and the meta data are flowing :-) Cheers!
